### PR TITLE
HIVE-21958 : The list of table expression in the inclusion and exclusion list should be separated by '|' instead of comma.

### PR DESCRIPTION
**Pull request by maheshk114**
…

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/DumpMetaData.java [hunk 1]
```diff
@@ -85,48 +85,27 @@ private void readReplScope(String line) throws IOException {
...
+        case 1:
+          LOG.info("Read ReplScope: Include table name list: {}.", value);
+          replScope.setIncludedTablePatterns(value);
+          break;
+        case 2:
+          LOG.info("Read ReplScope: Exclude table name list: {}.", value);
+          replScope.setExcludedTablePatterns(value);
+          break;
+        default:
+          LOG.warn("Found invalid info " + value + " in the repl dump file while reading ReplScope");
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/DumpMetaData.java]: Shall throw exception from here. There is no point in moving forward if this is incorrect.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java [hunk 2]
```diff
@@ -31,10 +29,10 @@
   private Pattern dbNamePattern;
 
   // Include and exclude table names/patterns exist only for REPL DUMP.
-  private List<String> includedTableNames;
-  private List<String> excludedTableNames;
-  private List<Pattern> includedTableNamePatterns;
-  private List<Pattern> excludedTableNamePatterns;
+  private String includedTableNames;
+  private String excludedTableNames;
+  private Pattern includedTableNamePatterns;
```
Comment (sankarh) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java]: Name can be includedTableNamesPattern and excludedTableNamesPattern as it is single pattern.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java [hunk 3]
```diff
@@ -93,26 +91,13 @@ public boolean tableIncludedInReplScope(final String tableName) {
     return (inTableIncludedList(tableName) && !inTableExcludedList(tableName));
   }
 
-  private List<Pattern> compilePatterns(List<String> patterns) {
-    if (patterns == null || patterns.isEmpty()) {
+  private Pattern compilePatterns(String pattern) {
```
Comment (sankarh) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java]: Name can be compilePattern as it is single pattern.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java [hunk 4]
```diff
@@ -313,8 +313,8 @@ public void testIncorrectTablePolicyInReplDump() throws Throwable {
     }
     Assert.assertTrue(failed);
 
-    // Invalid pattern where we didn't enclose table pattern within single or double quotes.
-    replPolicy = primaryDbName + ".[t1].[t2]";
+    // Invalid pattern where where table list is empty.
```
Comment (sankarh) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Type error. 2 "where"

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java [hunk 5]
```diff
@@ -313,8 +313,8 @@ public void testIncorrectTablePolicyInReplDump() throws Throwable {
...
+    replPolicy = primaryDbName + ".''.'t2'";
```
Comment (sankarh) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: We can treat empty list for dumping only functions. It was decided sometime earlier.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java [hunk 6]
```diff
@@ -263,12 +263,12 @@ public void testIncorrectTablePolicyInReplDump() throws Throwable {
...
-        primaryDbName + ".t1.t2", // Two explicit table names not allowed.
-        primaryDbName + ".['t1'].t2", // Table name and include list not allowed.
-        primaryDbName + ".t1.['t2']", // Table name and exclude list not allowed.
-        primaryDbName + ".[t1].t2", // Table name and include list not allowed.
-        primaryDbName + ".['t1+'].", // Abrubtly ended dot.
-        primaryDbName + "..[]" // Multiple dots
+        primaryDbName + ".t1.t2", // Didn't enclose table pattern within single quotes.
+        primaryDbName + ".'t1'.t2", // Table name and include list not allowed.
+        primaryDbName + ".t1.'t2'", // Table name and exclude list not allowed.
+        primaryDbName + ".'t1+'.", // Abrubtly ended dot.
```
Comment (sankarh) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: What is the behaviour for dbname.'t1'.'' and dbname.''? Empty include or exclude tables string.

Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Noticed it fails now. Do we need to support it to dump only functions?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/DumpMetaData.java (hunk 1 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/DumpMetaData.java]: done

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java (hunk 2 above)
Comment (maheshk114) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java]: done

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java (hunk 3 above)
Comment (maheshk114) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/common/repl/ReplScope.java]: done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 4 above)
Comment (maheshk114) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 5 above)
Comment (maheshk114) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: it's confusing ..if empty list ..that means include all ? if yes, then user has to give empty exclude list or exclude list as *  for dump only function ? i think for simplicity we should not allow this special case ..if user wants then they can use regex to exclude all tables

Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Yes, agreed.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 6 above)
Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: We can add the empty include/exclude list in this test case.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java (hunk 5 above)
Comment (sankarh) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: Check empty list for exclude list as well. Also add a test with more than 2 dots in the policy.

Comment (maheshk114) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestTableLevelReplicationScenarios.java]: done

**Pull request closed without merging**